fix: Bunch of fixes for revenue analytics managed viewsets

rafaeelaudibert opened the pull request:
- Include external tables
- Make sure paths are created
- Stop returning a bunch of `NULL`s for empty tables to make sure Arrow can materialize these views properly

rafaeelaudibert committed:
fix: Include external tables in managed viewset views

This is causing our views to fail materialization, oops

rafaeelaudibert commented on posthog/warehouse/models/datawarehouse_saved_query.py:
```diff
@@ -47,6 +47,30 @@ def validate_saved_query_name(value):
+    table_collector = S3TableVisitor()
+    table_collector.visit(node)
+
+    return list(table_collector.tables)
```
I extracted that function here because I wanna create the view with this already, so I don't have a view to call `s3_tables` on.

i can change that slightly to avoid this from being extracted, though, will update it accordingly

rafaeelaudibert committed:
fix: Make job error more readable

This helps debugging issues more easily

rafaeelaudibert committed:
fix: Stop crashing when parent doesn't have paths

This is all right, we don't need the parent to have a path

rafaeelaudibert committed:
fix: Create model path for viewset

rafaeelaudibert committed:
refactor: Go back to using `s3_tables`

Stop using the new function, just use the existing one

rafaeelaudibert committed:
fix: Stop creating `Nothing` values for materializatoin

These empty columns would have type `Nothing` when ran by Clickhouse which means they'd fail to be saved because they can't be converted to an Arrow type. Let's instead give them their proper types. We can't make them all just `UInt` because they're used in some places and they'd fail Clickhouse types validation - such as not letting you use UInt on a `toDatetime` call (even though the call would never happen because the result set is empty)

rafaeelaudibert committed:
fix: Guarantee `columns_with_type` is not none

rafaeelaudibert committed:
fix: Avoid returning hash/array

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

rafaeelaudibert committed:
tests: Remove incorrect test

The assumptions behind this test are wrong. We'll never have two saved views without any paths that will have their paths created after the fact. A view is always created with a path attached to it.

It's possible we update a view and fall into this trap, but this is tested below.

Follow-up to ad3924cd2aecd4ee6dfc4a16dcad7667e95ccd90

github-actions[bot] committed:
Update query snapshots

Gilbert09 reviewed (approved):
Just update the hogql func args and docstring, then this is good :)

rafaeelaudibert commented on posthog/hogql/ast.py:
```diff
@@ -831,16 +832,38 @@ class SelectQuery(Expr):
+        cls,
+        *,
+        columns: list[str] | None = None,
+        columns_with_type: dict[str, FieldOrTable] | None = None,
```
Yeah, but what about the places where we are NOT calling it like that, and we just have a list of strings? I'm trying to support the old usage without requiring them to pass in a very complicated `{key: UnknownDatabaseField for key in their_existing_list}`

rafaeelaudibert commented on posthog/hogql/ast.py:
```diff
@@ -831,16 +832,38 @@ class SelectQuery(Expr):
+                Alias(alias=column, expr=Constant(value=cast(DatabaseField, field).default_value()))
+                for (column, field) in columns_with_type.items()
+            ],
+            where=Constant(value=False),
```
Nope, it's still an empty query with no rows because of the `where`. It just goes from "Return an empty SelectQuery that evaulates to no rows with all columns typed as Unknown" to "Return an empty SelectQuery that evaulates to no rows with all columns typed as you wish"

rafaeelaudibert committed:
refactor: Simply `SelectQuery.empty`

Always receive a dict rather than a list of columns, this makes the structure much more type-safe

rafaeelaudibert committed:
refactor: Revert change to error

This might leak some context we don't wanna leak, so let's revert

rafaeelaudibert committed:
fix: Use new `SelectQuery.empty` format everywhere
